### ISSUE TITLE
files without extension in zip archive appear as directories in tr:unzip output

### ISSUE THREAD

**Issue by gimsieke** [transpect/unzip-extension#7]
This refers to the new (post-#1) output that contains c:directory elements.

If you extract attached archive [h.zip](https://github.com/transpect/unzip-extension/files/481345/h.zip) with the current tr:unzip, you will see the following output:

```
<c:files xmlns:c="http://www.w3.org/ns/xproc-step">
   <c:directory name="test">
      <c:directory name="README"/>
      <c:file name="README.md"/>
      <c:directory name="foo"/>
   </c:directory>
</c:files>
```

All entries below test/ are regular files. But only README.md is listed as a file. This is a severe bug that will, for example, report the mimetype file in EPUB, IDML, or ODT as a directory. I’m currently writing an adaption layer that will present the new output like the classic tr:unzip output, so that we don’t need to heavily adapt too many steps. But when converting a directory without files in it to the classic format, it will simply disappear.

**Comment by gimsieke** [comment]
If you fix this, please also restore the `/c:files/@xml:base` attribute (https://github.com/transpect/unzip-extension/issues/1#issuecomment-245586707).

**Comment by gimsieke** [comment]
It seems to be [coded](https://github.com/transpect/unzip-extension/blob/master/UnZip.java#L169) like that – the existence of a dot in a path component tells the step to identify this component as a file or as a directory. Seriously? If there is no better way to determine this from the zip entries list, the step shouldn’t output anything and instead rely on a recursive directory listing after unzipping. This will only report accurate results with respect to the zip contents if `overwrite=yes`, of course.

**Comment by gimsieke** [comment]
`overwrite=yes` is [not handled at all](https://github.com/transpect/unzip-extension/commit/16ac595d701ddb37ddb45c886b7a27d4b2277de9#diff-490133d5e8d954cab88ba4a4b30d634bR81). This makes a post-extraction recursive directory list pointless since there may be all kinds of leftovers from previous extractions. I’m quite dissatisfied that this half-baked implementation made its way into production, effectively blocking us from upgrading the Calabash distro.

**Comment by gimsieke** [comment]
If the information that is present in the archive’s list of entries does not allow to infer which files and folders will be created by unzipping, I don’t really see an alternative to recursive deletion (#2) before unzipping and a recursive directory list afterwards. 
Maybe we can address this by adding a default behavior that creates a subfolder of `dest-dir` (with a computed name) in which to unzip the archive. It will then try to recursively delete this subfolder first. This prevents people from deleting `/` or `C:/`. This default behavior should be made overridable. So we can introduce an option `safe=yes` that can be overridden by expert users who know what they are doing. 
If the unzip step calculates the actual destination folder below `dest-dir`, then it is important that it returns the URI of this folder in a response. It may return a `<c:zip-entries dest-dir="[user supplied OS path]" safe="[yes|no]" overwrite="[yes|no]" computed-dest-dir-uri="[file: URI]">…</c:zip-entries>` document that may contain a quite literal representation of the zip entries as reported by the zip inspection function. 
This raw unzip step should be wrapped in an XProc- (rather than Java-) implemented step that uses `/*/@computed-dest-dir-uri` for a recursive directory list and presents a hierarchical representation of this directory on one port (maybe the primary `result` port) and a backwards-compatible `flat` port (each file member of the archive represented by a `/c:files/c:file` entry).

**Comment by gimsieke** [comment]
This is my proposal:

– Revert to https://github.com/transpect/unzip-extension/commit/5438d48
– Make the java-implemented step available under a different namespace, for example `tr-internal=http://transpect.io/internal`, as `tr-internal:unzip`
– Make the currently imported step, that is imported by the canonical URI `http://transpect.io/calabash-extensions/unzip-extension/unzip-declaration.xpl`, into an XProc pipeline that wraps `tr-internal:unzip`. By default, it computes the safe subdir path below a given `dest-dir`, but people can opt to have `safe=no` (or `false`). 
– It deletes the target directory recursively and calls `tr-internal:unzip`.
– Then it creates a recursive directory listing.
– It makes this listing available on a port `list-including-directories`.
– We can add a port `zip-manifest` later. That port will produce a zip manifest that you can feed into `pxp:zip` or `tr:zip` to re-zip the directory. 
– On the primary `result` port, it either makes available the original output of the step, or it transforms the hierarchic directory list into a flattened variant. To my knowledge, there were no issues with the original output (except that it was not able to represent empty directories), therefore we may use the original output. 

The major drawback will be that the new `tr:unzip` relies on [xproc-util](https://github.com/transpect/xproc-util) and [xslt-util](https://github.com/transpect/xslt-util). These must be present in the project. If people don’t want these dependencies, they can still use the raw `tr-internal:unzip` step.

**Comment by gimsieke** [comment]
Or, other proposal. Revert to https://github.com/transpect/unzip-extension/commit/5438d48 and add a recursive-directory-list step if you need it. On 2nd thought, that wouldn’t solve the issue of accidental deletion. Ok, proposal withdrawn.

**Comment by mkraetke** [comment]
This problem was resolved by @gimsieke with commit 03bf82fadb980760d5b7ea5c8c66eefc9f454bf4. The step `tr:unzip` contains now a [secondary output port](https://github.com/transpect/unzip-extension/blob/master/unzip-declaration.xpl#L28-L31) which provides a recursive directory list.
